proposals/0189-sbpf-stricter-elf-headers.md

```diff
@@ -0,0 +1,105 @@
+---
+simd: '0189'
+title: SBPF stricter ELF headers
+authors:
+  - Alexander Meißner
+category: Standard
+type: Core
+status: Idea
+created: 2024-10-21
+feature: GJav1vwg2etvSWraPT96QvYuQJswJTJwtcyARrvkhuV9
+extends: SIMD-0178, SIMD-0179
+---
+
+## Summary
+
+Imposes more restrictions on what is expected of ELF headers.
+
+## Motivation
+
+After the removal of relocations in SIMD-0178 the ELF layout could be massively
+be simplified by constraining it to a strict subset of what ELF otherwise
+allows. Doing so not only reduces the complexity of validator implementations
+but also reduces the attack surface.
+
+## Alternatives Considered
+
+Moving away from ELF as a container format altogether. However this would only
+gain a very small file size advantage but otherwise loose all tooling
+compatibility.
+
+## New Terminology
+
+None.
+
+## Detailed Design
+
+The following must go into effect if and only if a program indicates the
+SBPF-version v3 or higher in its program header (see SIMD-0161).
+
+### File header
+
+The file size must not be less than `size_of::<Elf64Ehdr>()` (64 bytes),
+otherwise `ElfParserError::OutOfBounds` must be thrown.
+
+- `e_ident.ei_mag` must be `[0x7F, 0x45, 0x4C, 0x46]`
+- `e_ident.ei_class` must be `ELFCLASS64` (`0x02`)
+- `e_ident.ei_data` must be `ELFDATA2LSB` (`0x01`)
+- `e_ident.ei_version` must be `EV_CURRENT` (`0x01`)
+- `e_ident.ei_osabi` must be `ELFOSABI_NONE` (`0x00`)
+- `e_ident.ei_abiversion` must be `0x00`
+- `e_ident.ei_pad` must be `[0x00; 7]`
+- `e_type` must be `ET_REL` (`0x0001`)
+- `e_machine` must be `EM_BPF` (`0x00F7`)
+- `e_version` must be `EV_CURRENT` (`0x00000001`)
+- `e_entry` must point to a function start marker in the text section,
+see SIMD-0179 for details on how a function start is marked.
+- `e_phoff` must be `size_of::<Elf64Ehdr>()` (64 bytes)
+- `e_shoff` is not checked
+- `e_flags` see SIMD-0161
+- `e_ehsize` must be `size_of::<Elf64Ehdr>()` (64 bytes)
+- `e_phnum` must be `0x0002`
+- `e_phoff + e_phnum * size_of::<Elf64Phdr>()` must be less than the file size
```

Comment:
```suggestion
- `e_phoff + e_phnum * size_of::<Elf64Phdr>()` must be less than equal the file size
```

Doesn't our range look like this?
[`e_phoff`, `e_phoff + e_phnum * size_of::<Elf64Phdr>()`)
Meaning that `e_phoff + e_phnum * size_of::<Elf64Phdr>()` is exclusive and therefore should be okay to also be equal to the file size?

Verification of such an ELF will still fail later with `ElfParserError::InvalidProgramHeader`, but IMHO a header where `e_phoff + e_phnum * size_of::<Elf64Phdr>()` == `file.size()` holds should not fail with `ElfParserError::InvalidFileHeader` but `ElfParserError::InvalidProgramHeader`.

Reply:
I can adjust the comparison to be inclusive. But it is not really a useful case because if the program headers reach the end of the file, then their contents must be empty.